api/v1/mailables/MailableHandler.php

```diff
@@ -0,0 +1,145 @@
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class MailableHandler
+ * @ingroup api_v1_mailables
+ *
+ * @brief Base class to handle API requests for mailables.
+ */
+namespace PKP\api\v1\mailables;
+
+use APP\facades\Repo;
+use PKP\core\PKPApplication;
+use PKP\handler\APIHandler;
+use PKP\security\authorization\ContextRequiredPolicy;
+use PKP\security\authorization\PolicySet;
+use PKP\security\authorization\RoleBasedHandlerOperationPolicy;
+use PKP\security\Role;
+use Slim\Http\Request as SlimRequest;
+use Slim\Http\Response;
+
+class MailableHandler extends APIHandler
+{
+    /**
+     * @copydoc APIHandler::__construct()
+     */
+    public function __construct()
+    {
+        $this->_handlerPath = 'mailables';
+        $roles = [Role::ROLE_ID_SITE_ADMIN, Role::ROLE_ID_MANAGER];
+        $this->_endpoints = [
+            'GET' => [
+                [
+                    'pattern' => $this->getEndpointPattern(),
+                    'handler' => [$this, 'getMany'],
+                    'roles' => $roles,
+                ],
+                [
+                    'pattern' => $this->getEndpointPattern() . '/{className}',
+                    'handler' => [$this, 'get'],
+                    'roles' => $roles,
+                ],
+            ],
+            'PUT' => [
+                [
+                    'pattern' => $this->getEndpointPattern() . '/{className}',
+                    'handler' => [$this, 'edit'],
+                    'roles' => $roles,
+                ],
+            ],
+        ];
+        parent::__construct();
+    }
+
+    /**
+     * @copydoc PKPHandler::authorize
+     */
+    public function authorize($request, &$args, $roleAssignments)
+    {
+        $rolePolicy = new PolicySet(PolicySet::COMBINING_PERMIT_OVERRIDES);
+
+        // This endpoint is not available at the site-wide level
+        $this->addPolicy(new ContextRequiredPolicy($request));
+
+        foreach ($roleAssignments as $role => $operations) {
+            $rolePolicy->addPolicy(new RoleBasedHandlerOperationPolicy($request, $role, $operations));
+        }
+        $this->addPolicy($rolePolicy);
+
+        return parent::authorize($request, $args, $roleAssignments);
+    }
+
+    /**
+     * Get an array of mailables
+     */
+    public function getMany(SlimRequest $slimRequest, Response $response, array $args): Response
+    {
+        $mailables = Repo::mailable()->getMany($this->getRequest()->getContext()->getId(), $slimRequest->getQueryParam('searchPhrase'));
+
+        return $response->withJson($mailables, 200);
+    }
+
+    /**
+     * Get a single mailable
+     */
+    public function get(SlimRequest $slimRequest, Response $response, array $args): Response
+    {
+        $request = $this->getRequest();
+
+        $mailable = Repo::mailable()->getByMailableClassName($args['className'], true, $request->getContext()->getId());
```

Comment:
> This might lead to a fatal error in PHP 8.1 if $args['className'] is undefined.

Probably missing here something, but is it really possible to make such a request without `api.404.endpointNotFound` error?

Reply:
Good point! :+1: